specs/market-api.yaml

```diff
@@ -1125,29 +1189,94 @@ components:
         eventType:
           type: string
         eventDate:
-          type: string
-          format: date-time
+          $ref: 'common.yaml#/components/schemas/Timestamp'
+
+    PropertyQueryEvent:
+      allOf:
+        - $ref: '#/components/schemas/Event'
+        - type: object
+          required:
+            - propertyQuery
+          properties:
+            propertyQuery:
+              $ref: '#/components/schemas/PropertyQuery'
 
     ProposalEvent:
       allOf:
         - $ref: '#/components/schemas/Event'
         - type: object
+          required:
+            - proposal
           properties:
             proposal:
               $ref: '#/components/schemas/Proposal'
 
+    ProposalRejectedEvent:
+      allOf:
+        - $ref: '#/components/schemas/Event'
+        - type: object
+          required:
+            - proposalId
+          properties:
+            proposalId:
+              type: string
+              readOnly: true
+            reason:
+              $ref: 'common.yaml#/components/schemas/Reason'
+
     AgreementEvent:
       allOf:
         - $ref: '#/components/schemas/Event'
         - type: object
+          required:
+            - agreement
           properties:
             agreement:
               $ref: '#/components/schemas/Agreement'
 
-    PropertyQueryEvent:
+    # note this should be specified as a sub-class of Event, BUT
+    # a number of OpenAPI code generators fail on multistory inheritance hierarchies :(
+    AgreementOperationEvent:
+      type: object
+      required:
+        - eventType
+        - eventDate
+        - agreementId
+      discriminator:
+        propertyName: eventType
+      properties:
+        eventType:
+          type: string
+        eventDate:
+          $ref: 'common.yaml#/components/schemas/Timestamp'
+        agreementId:
+          type: string
+
+    AgreementApprovedEvent:
       allOf:
-        - $ref: '#/components/schemas/Event'
+        - $ref: '#/components/schemas/AgreementOperationEvent'
+        - type: object
+
+    AgreementRejectedEvent:
+      allOf:
+        - $ref: '#/components/schemas/AgreementOperationEvent'
         - type: object
           properties:
-            propertyQuery:
-              $ref: '#/components/schemas/PropertyQuery'
+            reason:
+              $ref: 'common.yaml#/components/schemas/Reason'
+
+    AgreementCancelledEvent:
+      allOf:
+        - $ref: '#/components/schemas/AgreementOperationEvent'
+        - type: object
+          properties:
+            reason:
+              $ref: 'common.yaml#/components/schemas/Reason'
+
+    AgreementTerminatedEvent:
```

Comment:
I'm wondering, maybe we should add information which side (Provider or Requestor) terminated Agreement. This could be important if breaking Agreement has some consequences negotiated within this Agreement. In current implementation we have no way do find out, who terminated, or am I wrong??

Reply:
moved to `AgreementTerminatedEvent`